packages/contracts/contracts/L1/Colosseum.sol

```diff
@@ -231,86 +313,114 @@ contract Colosseum is Initializable, Semver {
         );
 
-        bytes32[] memory dummyHashes;
-        if (_publicInput.txHashes.length < MAX_TXS) {
-            dummyHashes = Hashing.generateDummyHashes(
-                DUMMY_HASH,
-                MAX_TXS - _publicInput.txHashes.length
-            );
-        }
+        // TODO(pangssu): waiting for the new Verifier.sol to complete.
+        // require(ZK_VERIFIER.verify(_proof, _pair), "Colosseum: invalid proof");
 
-        // TODO(chokobole): check transaction root hash
-        // TODO(chokobole): use this public input hash require(_proof[4] == _publicInputHash);
-        bytes32 _publicInputHash = Hashing.hashPublicInput(
-            _srcOutputRootProof.stateRoot,
-            _publicInput,
-            CHAIN_ID,
-            dummyHashes
-        );
+        _validatePublicInput(_srcOutputRootProof, _publicInput, _rlps, _pair[4]);
+        challenge.outputRoot = dstOutputRoot;
+        challenge.current = msg.sender;
+        challenge.next = address(0);
 
-        require(ZK_VERIFIER.verify(_proof, _pair), "Colosseum: invalid proof");
+        // TODO(pangssu): call `createTransaction` to guardian
 
-        L2_ORACLE.deleteL2Outputs(challenge.outputIndex);
-        emit ProofCompleted(challengeId, challenge.outputIndex);
-        _closeChallenge(challengeId, challenge);
+        emit Proven(_outputIndex, dstOutputRoot);
     }
 
-    function asserterTimeout() external {
-        uint256 challengeId = latestChallengeId;
-        Types.Challenge storage challenge = challenges[challengeId];
+    /**
+     * @notice Closes the challenge because the challenger timed out.
+     *         The winner is the validator.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     */
+    function challengerTimeout(uint256 _outputIndex) external {
+        Types.Challenge storage challenge = challenges[_outputIndex];
         _validateTurn(challenge, ValidateTurnMode.NORMAL);
-        L2_ORACLE.deleteL2Outputs(challenge.outputIndex);
-        _closeChallenge(challengeId, challenge);
+        _closeChallenge(_outputIndex, challenge);
     }
 
-    function challengerTimeout(uint256 _challengeId) external {
-        Types.Challenge storage challenge = challenges[_challengeId];
-        _validateTurn(challenge, ValidateTurnMode.NORMAL);
-        _closeChallenge(_challengeId, challenge);
+    /**
+     * @notice Approve the challenge to replace output.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     */
+    function approveChallenge(uint256 _outputIndex) external {
+        require(msg.sender == GUARDIAN, "Colosseum: sender is not the guardian");
+
+        Types.Challenge storage challenge = challenges[_outputIndex];
+
+        require(
+            _challengeStatus(challenge) == ChallengeStatus.PROVEN,
+            "Colosseum: this challenge is not validated"
+        );
+
+        L2_ORACLE.replaceL2Output(_outputIndex, challenge.segments[0], challenge.current);
```

Comment:
The segmented output root is stored in `challenge.segments[0]`, which is guaranteed to be valid. In order to obtain the output root for replacement, aadditional arguments are required in `proveFault()`.

Reply:
I think we can use challenge.outputRoot which is stored in [line 344](https://github.com/kroma-network/kroma/pull/53/files#diff-77450785d1dafe0fddac317efe5f5151f0cd2d57405a72bccf457145c67e24a7R344).

Reply:
That output root does not seem to align with the submission interval (e.g., 1800, 3600, ...).

Reply:
oh I see it. You're right, thx!

Reply:
@seolaoh's point is exactly what I was going for. I tried to use the `challenge.outputRoot`, but missed it and couldn't change it.
If the challenge is proven, `challenge.outputRoot` is assigned.

Reply:
But that `challenge.outputRoot` is not the output root to be replaced.

Reply:
@chokobole you means, `dstOutputRoot` is not the output to replaced?

Reply:
Yes! for example, that output root is not output root of 3600 but somewhere in the middle between output root of 1800 and output root of 3600

Reply:
Oh, I see. Thanks!